contracts/peripheral/YieldDirectorV2.sol

```diff
@@ -0,0 +1,488 @@
+
+        sOHM = sOhm_;
+        gOHM = gOhm_;
+        staking = IStaking(staking_);
+    }
+
+    /************************
+     * Modifiers
+     ************************/
+    function isInvalidDeposit(uint256 amount_, address recipient_) internal view returns (bool) {
+        return depositDisabled || amount_ <= 0 || recipient_ == address(0);
+    }
+
+    function isInvalidUpdate(uint256 depositId_, uint256 amount_) internal view returns (bool) {
+        return depositDisabled || amount_ <= 0 || depositInfo[depositId_].depositor == address(0);
+    }
+
+    function isInvalidWithdrawal(uint256 amount_) internal view returns (bool) {
+        return withdrawDisabled || amount_ <= 0;
+    }
+
+    /************************
+     * Donor Functions
+     ************************/
+
+    /**
+        @notice Deposit gOHM, records sender address and assign rebases to recipient
+        @param amount_ Amount of gOHM debt issued from donor to recipient
+        @param recipient_ Address to direct staking yield and vault shares to
+    */
+    function deposit(uint256 amount_, address recipient_) external returns (uint256 depositId) {
+        depositId = _createDeposit(amount_, recipient_);
+
+        IERC20(gOHM).safeTransferFrom(msg.sender, address(this), amount_);
+    }
+
+    /**
+        @notice Deposit sOHM, wrap to gOHM, and records sender address and assign rebases to recipeint
+        @param amount_ Amount of sOHM debt issued from donor to recipient
+        @param recipient_ Address to direct staking yield and vault shares to
+    */
+    function depositSohm(uint256 amount_, address recipient_) external returns (uint256 depositId) {
+        uint256 gohmAmount = _toAgnostic(amount_);
+        depositId = _createDeposit(gohmAmount, recipient_);
+
+        IERC20(sOHM).safeTransferFrom(msg.sender, address(this), amount_);
+        IERC20(sOHM).approve(address(staking), amount_);
+        staking.wrap(address(this), amount_);
+    }
+
+    /**
+        @notice Deposit additional gOHM, and update deposit record
+        @param depositId_ Deposit ID to direct additional gOHM to
+        @param amount_ Amount of new gOHM debt issued from donor to recipient
+    */
+    function addToDeposit(uint256 depositId_, uint256 amount_) external {
+        _increaseDeposit(depositId_, amount_);
+
+        IERC20(gOHM).safeTransferFrom(msg.sender, address(this), amount_);
+    }
+
+    /**
+        @notice Deposit additional sOHM, wrap to gOHM, and update deposit record
+        @param depositId_ Deposit ID to direct additional gOHM to
+        @param amount_ Amount of new sOHM debt issued from donor to recipient
+    */
+    function addToSohmDeposit(uint256 depositId_, uint256 amount_) external {
+        uint256 gohmAmount = _toAgnostic(amount_);
+        _increaseDeposit(depositId_, gohmAmount);
+
+        IERC20(sOHM).safeTransferFrom(msg.sender, address(this), amount_);
+        IERC20(sOHM).approve(address(staking), amount_);
```

Comment:
Use safeApprove?  Also consider doing infinite approval in constructor to save gas